docs: update MNNVL user guide for single-annotation model (#587)
ai-dynamo/grove: docs/user-guide/auto-mnnvl.md

# Auto MNNVL (Multi-Node NVLink)

> Grove can automatically manage Multi-Node NVLink (MNNVL) acceleration for your GPU workloads. This guide explains how to enable, configure, and manage the auto MNNVL feature using the `grove.io/mnnvl-group` annotation.

Reviewer: ```suggestion
Grove can automatically manage Multi-Node NVLink (MNNVL) setup for your GPU workloads. This guide explains how to enable, configure, and manage the auto MNNVL feature using the `grove.io/mnnvl-group` annotation. It is Grove's intention to keep the core API vendor agnostic which is why MNNVL support is handled via an annotation.
```

Author: fix

Without the auto MNNVL feature, you must manually create ComputeDomain resources and wire up `resourceClaims` in your pod specs. With auto MNNVL enabled, Grove handles this automatically:

- Detects GPU containers in your PodCliqueSet (PCS)
- Creates one ComputeDomain per MNNVL group per PCS replica
- Injects resource claim references into enrolled PodClique (PCLQ) pod specs
- Manages the full ComputeDomain lifecycle (creation, scaling, deletion)

You control MNNVL participation with a single annotation, `grove.io/mnnvl-group`, which can be placed at any level of the PCS hierarchy:

| Annotation value | Meaning |
|---|---|
| A group name (e.g., `"my-group"`, `"workers"`) | Opt in to MNNVL. PodCliques with the same group name share a ComputeDomain per replica. |
| `"none"` | Explicit opt-out. Overrides a parent layer's group assignment. |
| Absent | Inherit from the parent layer. If no parent sets it, no MNNVL. |

## Prerequisites and Constraints

Before enabling auto MNNVL, ensure your cluster meets the following requirements:

> 1. **NVIDIA GPUs with MNNVL support** on the nodes that will run MNNVL workloads.

Reviewer: ```suggestion
1. **NVIDIA GPUs with MNNVL support** on the nodes that will run MNNVL workloads e.g. GB200 and GB300 NVL72 systems.
```

Author: Done

When the feature is disabled (`autoMNNVLEnabled: false`), any PCS that requests MNNVL (i.e., `grove.io/mnnvl-group` set to a group name) is rejected at admission time.

## How It Works

You opt in to MNNVL by adding the `grove.io/mnnvl-group` annotation at any layer of the PCS hierarchy — PCS, PodCliqueScalingGroup (PCSG), or PodClique (PCLQ). The annotation value is a group name — PodCliques sharing the same group name share a single ComputeDomain per replica. The ComputeDomains and all necessary resource claims are managed entirely by Grove. Your pods automatically join the MNNVL fabric when scheduled.

The annotation propagates downward through the hierarchy: **PCS → PCSG → PCLQ**. A lower layer overrides a higher layer, so you can set a default at the PCS level and override it on specific PodCliques or PodCliqueScalingGroups.

| PCS annotation | PCLQ annotation | Effective PCLQ group |
|---|---|---|
| `"my-group"` | Absent | `"my-group"` (inherited) |
| `"my-group"` | `"other-group"` | `"other-group"` (overridden) |
| `"my-group"` | `"none"` | No MNNVL (opted out) |
| Absent | `"my-group"` | `"my-group"` (PCLQ-level only) |

> ### Non-GPU PodCliques

Reviewer: When we say “non-GPU PodClique,” do we specifically mean any PodClique that does not request `nvidia.com/gpu`? I want to confirm that PodCliques requesting non-NVIDIA GPUs would also be ignored by MNNVL rather than treated as eligible GPU PodCliques. If this is correct we should specify that its non-nvidia GPU PodClique.

It may also be worth adding guidance for mixed GPU clusters. For example, in a cluster with both H100s and NVL72-capable GPUs, PodCliques that should not participate in MNNVL should either avoid inheriting an MNNVL group or explicitly opt out with `grove.io/mnnvl-group: "none"`. My understanding is that if the DRA driver is installed across all NVIDIA GPUs, including the non-GB200/GB300 cases in a ComputeDomain is technically safe because the driver is a no-op there, but for readability and intent it would be clearer to explicitly exclude those PodCliques from the MNNVL group so I think would be good to explain this.

Author: Good catch — updated to 'non-NVIDIA-GPU' and added mixed GPU cluster guidance

                    nvidia.com/gpu: "8"
```

> This creates ComputeDomains `my-workload-0-my-group` and `my-workload-1-my-group`, for all PodCliques.

Reviewer: ```suggestion
In this example, spec.replicas: 2 produces two PCS replicas, my-workload-0 and my-workload-1. The operator creates one ComputeDomain per PCS replica by appending the MNNVL group name, resulting in my-workload-0-my-group and my-workload-1-my-group. Because no per-PodClique MNNVL group overrides are specified, all GPU PodCliques within each PCS replica use the PCS-level group and share that replica’s ComputeDomain.
```

Author: fix

### Multiple MNNVL Groups

> Assign different PodCliques to separate groups. Each group gets its own ComputeDomain per replica:

Reviewer: ```suggestion
Assign different PodCliques to separate groups. Each group gets its own ComputeDomain per PodCliqueSet replica:
```

Author: fix

```yaml
apiVersion: grove.io/v1alpha1
kind: PodCliqueSet
metadata:

>   name: training-job

Reviewer: Since we dont really have training job support yet lets just call this inference-deployment, i guess you could imagine a situation where workers are the main model replicas, encoders are the vision encoders, and then lets change param-servers to frontend, will add some suggestions to make these changes.

```suggestion
  name: inference-deployment
```

Author: Done — renamed to `inference-deployment` with `frontends`

spec:
  replicas: 2
  template:
    cliques:
      - name: workers
        annotations:
          grove.io/mnnvl-group: "workers"
        spec:
          replicas: 1
          podSpec:
            containers:
              - name: worker
                image: my-model:latest
                resources:
                  limits:
                    nvidia.com/gpu: "8"
      - name: encoders
        annotations:
          grove.io/mnnvl-group: "encoders"
        spec:
          replicas: 1
          podSpec:
            containers:
              - name: encoder
                image: my-encoder:latest
                resources:
                  limits:
                    nvidia.com/gpu: "4"

>       - name: param-servers

Reviewer: ```suggestion
      - name: frontends
```

        spec:
          replicas: 1
          podSpec:
            containers:

>               - name: ps

Reviewer: ```suggestion
              - name: frontend
```

>                 image: my-ps:latest

Reviewer: ```suggestion
                image: my-frontend:latest
```

                resources:
                  limits:
                    cpu: "4"
```

> Result per replica:

Reviewer: ```suggestion
Result per PCS replica:
```

- `workers` → group `"workers"` → ComputeDomain `training-job-0-workers` / `training-job-1-workers`
- `encoders` → group `"encoders"` → ComputeDomain `training-job-0-encoders` / `training-job-1-encoders`

> - `param-servers` → no annotation → no MNNVL

Reviewer: ```suggestion
- `frontends` → non GPU PodClique → no MNNVL
```

                resources:
                  limits:
                    nvidia.com/gpu: "8"
      - name: gpu-c
        spec:
          replicas: 1
          podSpec:
            containers:
              - name: c
                image: my-model:latest
                resources:
                  limits:
                    nvidia.com/gpu: "8"
      - name: gpu-d
        annotations:
          grove.io/mnnvl-group: "none"
        spec:
          replicas: 1
          podSpec:
            containers:
              - name: d
                image: my-model:latest
                resources:
                  limits:
                    nvidia.com/gpu: "8"
```

Resolution for each PodClique:

| PodClique | PCLQ annotation | PCSG annotation | PCS annotation | Effective group |
|---|---|---|---|---|
| `gpu-a` | Absent | `"none"` | `"my-group"` | No MNNVL — PCSG opts out |
| `gpu-b` | `"my-group"` | `"none"` | `"my-group"` | `"my-group"` — PCLQ overrides PCSG back into the PCS group |
| `gpu-c` | Absent | Absent | `"my-group"` | `"my-group"` — inherited through PCSG from PCS |
| `gpu-d` | `"none"` | Absent | `"my-group"` | No MNNVL — PCLQ opts out |

> ### Group Name Requirements

Reviewer: I'm thinking Group Name Requirements and ComputeDomain Naming should go above the usage examples? The remaining stuff below seems more tied to the example above so i think fine to keep it as is, what do you think?

Author: Agreed — moved both sections above Usage Examples
